tools/shape_tool/ShapeWindow.cpp

```diff
@@ -0,0 +1,162 @@
+#include "../../../../lib/ui3/elements/TileShapeGraphic.cpp"
+
+const array<int> SHAPE_MAP = {
+	 8, 16,  0,  0,
+	 7, 15,  0,  0,
+	11,  3, 20, 17,
+	12,  4, 19, 18,
+	10,  9,  1,  2,
+	 6,  5, 13, 14
+};
+
+class ShapeWindow : Window
+{
+
+	private Container@ shape_container;
+	private Button@ custom_button;
+
+	private int _tile_shape = 0;
+	private int _selection_index = 2;
+
+	private bool _selecting_tiles = false;
+	private bool _using_custom_shape = false;
+
+	private TileShapes@ _selected_shape;
+
+	int tile_shape
+	{
+		get const { return _tile_shape; }
+		set
+		{
+			if(value < 0 or value > 20) return;
+			int index = SHAPE_MAP.find(value);
+			cast<Button>(shape_container.get_child(index)).selected = true;
+		}
+	}
+
+	bool selecting_tiles { get const { return _selecting_tiles; } }
+	bool using_custom_shape { get const { return _using_custom_shape; } }
+
+	ShapeWindow(UI@ ui)
+	{
+		super(ui, "Shape", false);
+
+		@layout = FlowLayout(ui, FlowDirection::Row);
+
+		@shape_container = Container(ui);
+
+		GridLayout@ grid = GridLayout(ui, 4);
+		grid.row_spacing = 0;
+		grid.column_spacing = 0;
+		@shape_container.layout = grid;
+
+		ButtonGroup@ button_group = ButtonGroup(ui, false);
+		button_group.select.on(EventCallback(on_button_select));
+
+		for(int i=0; i<24; ++i)
+		{
+			int tile_shape = SHAPE_MAP[i];
+			TileShapeGraphic@ t = TileShapeGraphic(ui, tile_shape);
+
+			Button@ b = Button(ui, t);
+			b.name = "" + i;
+			b.width = 48;
+			b.height = 48;
+			b.selectable = true;
+
+			if(i == 2) 
+				b.selected = true;
+
+			button_group.add(b);
+			shape_container.add_child(b);
+		}
+
+		shape_container.fit_to_contents();
+
+		Container@ custom_shape_container = Container(ui);
+		@custom_shape_container.layout = FlowLayout(ui, FlowDirection::Row);
+
+		Label@ custom_label = Label(ui, "Clone");
+		custom_label.width = 2 * 48 - 2 * ui.style.spacing;
+		custom_label.height = 48;
+		custom_label.align_h = GraphicAlign::Centre;
+		custom_label.align_v = GraphicAlign::Middle;
+
+		Button@ select_button = Button(ui, "X");
+		select_button.name = "select";
+		select_button.width = 48;
+		select_button.height = 48;
+		select_button.selectable = true;
+		@select_button.tooltip = PopupOptions(ui, "Select a region", false, PopupPosition::Below);
+
+		@custom_button = Button(ui, "O");
+		custom_button.name = "custom";
+		custom_button.width = 48;
+		custom_button.height = 48;
+		custom_button.selectable = true;
+		custom_button.disabled = true;
+		@custom_button.tooltip = PopupOptions(ui, "Paste tiles", false, PopupPosition::Below);
+
+		button_group.add(select_button);
+		button_group.add(custom_button);
+
+		custom_shape_container.add_child(custom_label);
+		custom_shape_container.add_child(select_button);
+		custom_shape_container.add_child(custom_button);
```

Comment:
Windows do have a built button bar, so might be better to use instead of creating your own container.

Reply:
Changed the buttons a little and used the built-in bar, so now it looks like this:

![image](https://github.com/cmann1/adv-tools/assets/16232336/97738dca-f3d9-4f76-951d-360c7ef98e80)

I had originally wanted both buttons to be next to each other, right aligned. But calling `add_button_right` twice stacked them on top of each other. I think this looks fine as it is though.